### PR TITLE
adding specs for SQL namespace

### PR DESCRIPTION
**Pull request by Tokesh**
### Description
adding sql query specification

### Issues Resolved
[#235]

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
Awesome! Please write some tests for it? See https://github.com/opensearch-project/opensearch-api-specification/blob/main/SPECIFICATION_TESTING.md and https://github.com/opensearch-project/opensearch-api-specification/tree/main/tests.

**Comment by Tokesh** [comment]
> Awesome! Please write some tests for it? See https://github.com/opensearch-project/opensearch-api-specification/blob/main/SPECIFICATION_TESTING.md and https://github.com/opensearch-project/opensearch-api-specification/tree/main/tests.

Hello! Yes, i'll add test a bit later. I'm just getting used to OpenAPI.

**Comment by dblock** [comment]
Want to get this to a mergeable state? It doesn't have to be full coverage of the SQL plugin, can start with just the basics and make more PRs afterwards!

**Comment by Tokesh** [comment]
> Want to get this to a mergeable state? It doesn't have to be full coverage of the SQL plugin, can start with just the basics and make more PRs afterwards!

Hi! I wrote it, I think I corrected it. At least you can start watching. The only thing is that the tests ran locally for me and passed the test, but in Github Actions they did not pass the test;
image

File: spec/namespaces/sql.yaml [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+  title: OpenSearch SQL API
+  description: OpenSearch SQL API
+  version: 1.0.0
+paths:
+  /_plugins/_sql:
+    post:
+      operationId: sql.query.0
+      x-operation-group: sql.query
+      x-version-added: '1.0'
+      description: Sends an SQL/PPL query to the SQL plugin.
```
Comment (dblock) [inline spec/namespaces/sql.yaml:12]: Send a ...

File: spec/namespaces/sql.yaml [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        $ref: '#/components/requestBodies/sql.explain'
+      responses:
+        '200':
+          $ref: '#/components/responses/sql.explain@200'
+  /_plugins/_sql/close:
+    post:
+      operationId: sql.close.0
+      x-operation-group: sql.close
+      x-version-added: '1.0'
+      description: To explicitly clear the cursor context.
```
Comment (dblock) [inline spec/namespaces/sql.yaml:44]: Clears the cursor context.

File: spec/namespaces/sql.yaml [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        $ref: '#/components/requestBodies/sql.close'
+      responses:
+        '200':
+          $ref: '#/components/responses/sql.close@200'
+  /_plugins/_sql/stats:
+    get:
+      operationId: sql.get_stats.0
+      x-operation-group: sql.get_stats
+      x-version-added: '1.0'
+      description: By a stats endpoint, you are able to collect metrics for the plugin within the interval.
```
Comment (dblock) [inline spec/namespaces/sql.yaml:60]: Collect metrics ...

**Review by dblock — CHANGES_REQUESTED** [review]
Close. Let's get CI to green.

1. For the linter run `npm run lint--fix` and then correct any remaining lints till it's clean manually. Some words may need to be added to .cspell.
2. One of the tests did fail in CI. I don't see an error why, so this will need a bit more digging, maybe commit `--verbose` into CI and we can see what's up. It should at least show an error, which is probably a bug in the test framework that I can help fix.

**Comment by nhtruong** [comment]
@dblock [That](https://github.com/opensearch-project/opensearch-api-specification/actions/runs/10080305941/job/27869725926?pr=379#step:6:83)'s very unusual for the framework to obscure an error not related to the testing itself. This should not have been handled by the framework and should crash it for the error trace to be shown.

**Comment by dblock** [comment]
> @dblock [That](https://github.com/opensearch-project/opensearch-api-specification/actions/runs/10080305941/job/27869725926?pr=379#step:6:83)'s very unusual for the framework to obscure an error not related to the testing itself. This should not have been handled by the framework and should crash it for the error trace to be shown.

Yes, 100%. I can look into it.

**Comment by nhtruong** [comment]
Created an issue: https://github.com/opensearch-project/opensearch-api-specification/issues/449

**Comment by Tokesh** [comment]
Probably i found misstake
I'm waiting for the cursor to receive, but it doesn't come with the API
I will test and see

**Comment by Tokesh** [comment]
I'm so happy! You can check it now!
@dblock @nhtruong

**Review by nhtruong — DISMISSED** [review]
This is great! Thank you!

**Comment by nhtruong** [comment]
Cant merged. Blocked by https://github.com/opensearch-project/opensearch-api-specification/pull/379#pullrequestreview-2196502955 @dblock

File: CHANGELOG.md [hunk 4]
```diff
@@ -35,6 +35,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - Added AjvErrorsParser to print more informative error messages ([#364](https://github.com/opensearch-project/opensearch-api-specification/issues/364))
 - Added JsonSchemaValidator, a wrapper for AJV ([#364](https://github.com/opensearch-project/opensearch-api-specification/issues/364))
 - Added support for `application/cbor` responses ([#371](https://github.com/opensearch-project/opensearch-api-specification/pull/371))
+- Added tests for SQL namespace ([#379](https://github.com/opensearch-project/opensearch-api-specification/pull/379))
```
Comment (dblock) [inline CHANGELOG.md:38]: Actually the add is the spec.

Let's say "Added `/_plugins/_sql`, `close` and `stats`"?

File: tests/sql/close.yaml [hunk 5]
```diff
@@ -0,0 +1,53 @@
+$schema: ../../json_schemas/test_story.schema.yaml
+
+description: Test to explicitly clear the cursor context
```
Comment (dblock) [inline tests/sql/close.yaml:3]: Add a period at the end.

File: tests/sql/close.yaml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        - {author: Harper Lee, title: To Kill a Mockingbird, year: 1960}
+        - {create: {_index: books, _id: book_1392215}}
+        - {author: Elizabeth Rudnick, title: Beauty and the Beast, year: 1991}
+  - synopsis: Refresh the index.
+    path: /{index}/_refresh
+    method: POST
+    parameters:
+      index: books
+    response:
+      status: 200
```
Comment (dblock) [inline tests/sql/close.yaml:33]: The above should be moved to a prologue because they aren't exercising the SQL API namespace.

**Review by dblock — CHANGES_REQUESTED** [review]
Couple more. I think we should stay disciplined about prologue, sorry to be a hassle.

**Comment by dblock** [comment]
> Probably i found misstake I'm waiting for the cursor to receive, but it doesn't come with the API I will test and see

It did raise that bug of reporting errors, so thank you! Fix in https://github.com/opensearch-project/opensearch-api-specification/pull/450.

File: tests/sql/close.yaml (hunk 5 above)
Comment (nhtruong) [reply tests/sql/close.yaml:3]: Time for another lint!

**Comment by Tokesh** [comment]
> Couple more. I think we should stay disciplined about prologue, sorry to be a hassle.
> 
> Also see if you can exercise the parameters like `sanitize` that aren't.

I found `sanitize` in documentation here:
image
And rechecked on local machine

**Review by nhtruong — APPROVED** [review]

**Comment by dblock** [comment]
> > Also see if you can exercise the parameters like `sanitize` that aren't.
> 
> I found `sanitize` in documentation here: image And rechecked on local machine

Yes, I meant adding a test that actually uses it.

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
Merged this! @Tokesh if you want to remove the changelog line that says "Added tests for ... SQL", add some tests for `?sanitize` and whatever other parts of the spec you declared but didn't test in a next PR, that'd be great!

File: tests/sql/close.yaml (hunk 5 above)
Comment (dblock) [reply tests/sql/close.yaml:3]: We have https://github.com/opensearch-project/opensearch-api-specification/issues/403, I'll add a note on tests.

**Pull request merged**